### PR TITLE
Update url validity check with better messages and tag priority

### PR DESCRIPTION
**Pull request by tfoote**
This is capturing the requirements documented in #20286 more clearly in the unit tests.

### CODE REVIEW

**Review by nuclearsandwich — DISMISSED** [review]

**Comment by nuclearsandwich** [comment]
Per the ongoing discussion in #20309 I've updated this PR to allow source entries to contain tags when test_commits and test_pull_requests are both explicitly false. I've also added changes based on https://github.com/ros/rosdistro/pull/20261 (being sure to give co-authorship to @k-okada) which allow commits to be tested as well.

File: test/test_url_validity.py [hunk 1]
```diff
@@ -76,50 +77,77 @@ def detect_lines(diffstr):
...
+    if not version:
+        # If the above passed assume the default exists
+        return (True, '')
+
+    if 'refs/heads/%s' % version in output:
+        return (True, '')
+
+    if commits_valid:
+        try:
+            tmpdir = tempfile.mkdtemp()
```
Comment (tfoote) [inline test/test_url_validity.py]: This tmpdir needs to get cleaned up.

File: test/test_url_validity.py [hunk 2]
```diff
@@ -76,50 +77,77 @@ def detect_lines(diffstr):
...
-def check_source_repo_entry_for_errors(source, tags_valid=False):
+    try:
+        output = subprocess.check_output(cmd).decode('utf-8')
+    except subprocess.CalledProcessError as ex:
+        return (False, 'subprocess call %s failed: %s' % (cmd, ex))
+    if not version:
+        # If the above passed assume the default exists
+        return (True, '')
+
+    if 'refs/heads/%s' % version in output:
```
Comment (tfoote) [inline test/test_url_validity.py]: This is technically incorrect to test before the commits. However testing the commits is very expensive. I'd suggest that we only test the commits if the `version` matches the appropriate regex of 40 characters and `[0-9a-f]`

**Review by tfoote — COMMENTED** [review]
Sorry I got half way through this and my review comments dissappeared. Apparently I can only comment on this now too. Edit: Oh because it's my PR originally ;-)

File: test/test_url_validity.py (hunk 1 above)
Comment (nuclearsandwich) [reply test/test_url_validity.py]: I took a quick look in the tempfile library docs but on the off chance it's there and I didn't find it, is there a context manager based way to mange temp directories a la

```
with tempfile.mkdtemp() as tmpdir
  ....
```

File: test/test_url_validity.py (hunk 2 above)
Comment (nuclearsandwich) [reply test/test_url_validity.py]: 610d309

File: test/test_url_validity.py (hunk 1 above)
Comment (nuclearsandwich) [reply test/test_url_validity.py]: e8dc20a

**Review by tfoote — COMMENTED** [review]
I added a commit updating the order of checking.

**Comment by nuclearsandwich** [comment]
It looks like Tully and I both think this is ready. Anyone else want to give it a :mag: ?

**Comment by k-okada** [comment]
thanks!, LGTM

**Pull request merged**